proj w/o geo

michaelkirk opened the pull request:
It seems like the potential users of proj would include people who don't need/want to opt into the whole `geo` ecosystem.

What would you think about having the proj interface be trait-based and having the geo-integration be a feature?

If this seems like a reasonable direction, before merging I'd like to update the docs to show:

1. how you can use this with the greater `geo` ecosystem
2. how you can `impl` your own point type if you're not using geo.
3. how you can use the built-in `(Float, Float)` impl

WDYT @urschrei?

michaelkirk commented on src/geo_types.rs:
````diff
@@ -0,0 +1,53 @@
+/// assert_approx_eq!(result.x, 1450880.29f64, 1.0e-2);
+/// assert_approx_eq!(result.y, 1141263.01f64, 1.0e-2);
+/// ```
+impl<T: crate::proj::CoordinateType> crate::Point<T> for geo_types::Coordinate<T> {
````
Note - this PR would also let you do Proj stuff with Coords, not just points.

michaelkirk replied:
Now that I'm typing it out - we could take this even further and have a `Projectable` trait or somesuch that just required implementing a Iter<Item=proj::Point<T>>, then we could get easy projection of all the geo-type's. 🤔

michaelkirk replied:
I'm not going to pursue the Iter scenario in this PR, but would like to do it in some followup work.

michaelkirk commented:
We could make the `geo-types` feature a default feature to make this a non-breaking change.

urschrei commented:
I'm in favour of this so long as we write clear documentation around it. In general, I find it tricky to present feature interaction in a clear and concise way, but we needn't let that stop us…

michaelkirk commented on README.md:
````diff
@@ -22,10 +22,7 @@ By default, this crate depends on a pre-built `libproj`, so PROJ v7.1.x must be
 
 ## Convert from [NAD 83 US Survey Feet](https://epsg.io/2230) to [NAD 83 Meters](https://epsg.io/26946) Using EPSG Codes
 ```rust
````
The README contains lots of snippets from the primary docs. Since it can be tedious and error prone to copy changes back and forth between them, I was hoping we could review just the primary docs, and once they're in good shape I can update the README before a final review.

michaelkirk reviewed (commented):
Ok, with the caveat of the README (see comment) I think this is ready for review. Can you take a look @urschrei?

urschrei reviewed (approved):
This lgtm, let's merge!

urschrei commented:
I think this probably just needs a rebase now that we've switched to GH Actions?

michaelkirk commented:
Ok! Rebased and addressed conflicts in the docs. 

Note that I'm (almost) wholesale copying the contents of README.md from the doc comments in src/lib.rs - the only divergence is the GH actions badge.

michaelkirk commented:
bors try

michaelkirk commented on Cargo.toml:
```diff
@@ -14,7 +14,7 @@ edition = "2018"
 [dependencies]
 proj-sys = { version = "0.18.3", path = "proj-sys" }
-geo-types ="0.6.0"
+geo-types = { version = "0.6", optional = true }
```
I snuck this in - there wasn't a reason for us to peg the patch version of geo-types was there?

michaelkirk commented:
bors try

I'm not sure why - but [bors crashed](https://app.bors.tech/repositories/42/log)... let's try again.

michaelkirk commented:
bors r=urschrei

michaelkirk commented:
bors cancel

michaelkirk commented:
🤦 booooorrrsss

Oh well, I just wanted to fix a little typo in the README and squash the commits, but bors will do what bors wants.

michaelkirk commented:
Oh, actually that was my bad. I should have said `bors r-` instead of `bors cancel`. Sorry for doubting you bors 😬.

urschrei commented:
We have always known bots and people couldn't live in harmony. It's been a good run. (Thank you for doing this!)

michaelkirk commented:
wth bors

michaelkirk commented:
![image](https://user-images.githubusercontent.com/217057/103804296-b7496b00-5006-11eb-8423-b7122c434cb2.png)
That eye. So piercing.

urschrei commented:
I feel judged.
